### PR TITLE
Improve media session compliance

### PR DESCRIPTION
**Pull request by morckx**
Can be tested with:
https://github.com/morckx/android-media-controller/tree/dont-tint-mediaitem-icons

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -71,6 +71,7 @@ dependencies {
     implementation 'jp.wasabeef:picasso-transformations:2.2.1'
     implementation 'com.google.android.exoplayer:exoplayer:2.9.6'
     implementation 'com.github.rustamg:file-dialogs:1.0'
+    implementation group: 'info.debatty', name: 'java-string-similarity', version: '1.2.1'
```
Comment (werman) [inline app/build.gradle]: Small nitpick: better to be consistent in the naming with other items.

File: app/src/main/java/net/programmierecke/radiodroid2/MediaSessionCallback.java [hunk 2]
```diff
@@ -101,4 +104,16 @@ public void onPlayFromMediaId(String mediaId, Bundle extras) {
             bm.sendBroadcast(intent);
         }
     }
+
+    @Override
+    public void onPlayFromSearch(String query, Bundle extras) {
+        DataRadioStation station = ((RadioDroidApp) context.getApplicationContext()).getFavouriteManager().getBestNameMatch(query);
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/MediaSessionCallback.java]: I think that maybe history also should be included.

File: app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java [hunk 3]
```diff
@@ -92,6 +93,23 @@ public void move(int fromPos, int toPos) {
         Collections.rotate(listStations.subList(Math.min(fromPos, toPos), Math.max(fromPos, toPos) + 1), Integer.signum(fromPos - toPos));
     }
 
+    public @Nullable DataRadioStation getBestNameMatch(String query) {
+        DataRadioStation bestStation = null;
+        query = query.toUpperCase();
+        double smallesDistance = Double.MAX_VALUE;
+
+        for (DataRadioStation station : listStations) {
+            Cosine distMeasure = new Cosine();
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java]: Probably it shouldn't be created inside the loop.

File: app/build.gradle (hunk 1 above)
Comment (morckx) [reply app/build.gradle]: Oops, yes.

File: app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java (hunk 3 above)
Comment (morckx) [reply app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java]: Unfortunately at least the measures it tested this with required to be renewed every time, but I'll check if cosine (which seems to work quite well for keyboard input)  is affected as well.

Comment (werman) [reply app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java]: Ah, I love such library design.

Comment (werman) [reply app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java]: If it true it worth adding a comment.

**Comment by morckx** [comment]
> I have mostly just glanced over and didn't actually test this.
> With these changes we should be closer to android auto support. However I'm not sure that searching in favourites/history would be enough for that...

Why not? We could also fall back to server search, but I think it's even better to only search in favorites and history. What I'd love to have is google assistant support, wich currently only works with tune in.

**Comment by werman** [comment]
> > I have mostly just glanced over and didn't actually test this.
> > With these changes we should be closer to android auto support. However I'm not sure that searching in favourites/history would be enough for that...
> 
> Why not? We could also fall back to server search, but I think it's even better to only search in favorites and history. What I'd love to have is google assistant support, wich currently only works with tune in.

It probably would be enough but from my pov it would be much **better** to have fallback to fuzzy search across all servers which currently cannot be done since server doesn't do fuzzy search and on client we don't have full database. However I don't actually know how real users will use it: would they search new stations with it or only ones they used.

**Comment by morckx** [comment]
> It probably would be enough but from my pov it would be much **better** to have fallback to fuzzy search across all servers which currently cannot be done since server doesn't do fuzzy search and on client we don't have full database. 

I think a substring keyword search would come quite close to a fuzzy search.

> However I don't actually know how real users will use it: would they search new stations with it or only the ones they used.

Can we activate android auto (the launcher icon?) again for a beta release? 
Maybe @beccon4 would be willing to test it.

Personally, for voice searches, I'd prefer to only search in the favorites, but that could be made configurable.

> 
> P.S. And I use neither android auto nor any kind of voice assistant so I think I'm too out of touch with this.

I don't use android auto either, but would like to use google assistant as remote control for the kitchen radio application.

**Comment by morckx** [comment]
> Can we activate android auto (the launcher icon?) again for a beta release?

OK, let's do it in a second step, if it's possible.

File: app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java (hunk 3 above)
Comment (morckx) [reply app/src/main/java/net/programmierecke/radiodroid2/StationSaveManager.java]: OK, for Cosine it works inside the loop. I've added a warning comment for other measures.

**Comment by morckx** [comment]
> It probably would be enough but from my pov it would be much better to have fallback to fuzzy search across all servers which currently cannot be done since server doesn't do fuzzy search and on client we 

I've come to the conclusion that it doesn't make sense to put too much work into fuzzy search for this purpose, because as a result you don't get a list of hits, but the best match is played directly. 

What I imagine is a settings option: fall back to server search if no exact hits  (meaning the best matching station name string contains the query string) in favourites or history.

**Comment by werman** [comment]
> Can we activate android auto (the launcher icon?) again for a beta release?

I don't think so, we'd need to check if we are compliant with the android auto's requirements, the app must pass all of them to be published. E.g. in `onPlayFromSearch` we should handle what exactly the query is, see https://developer.android.com/training/cars/media#parse_voice

> I think a substring keyword search would come quite close to a fuzzy search.

I think in the end it just requires some testing because I don't know if voice recognition will pass good enough results for just substring search - which I highly doubt. However it's a mater for other time.

**Comment by morckx** [comment]
> > Can we activate android auto (the launcher icon?) again for a beta release?
> 
> I don't think so, we'd need to check if we are compliant with the android auto's requirements, the app must pass all of them to be published. E.g. in `onPlayFromSearch` we should handle what exactly the query is, see https://developer.android.com/training/cars/media#parse_voice

Thanks, that's very interesting! There are indeed some things left we'd need to take care of. Not very nice without having a corresponding car stereo to test and benefit.

> 
> > I think a substring keyword search would come quite close to a fuzzy search.
> 
> I think in the end it just requires some testing because I don't know if voice recognition will pass good enough results for just substring search - which I highly doubt. However it's a mater for other time.

I wouldn't be too pessimistic here. What we offer with this PR is already on par with what YouTube does with voice search results: extremely recall oriented opportunistic search, with a high chance of false positives. So we can't learn from YouTube, but maybe from TuneIn.

**Comment by werman** [comment]
> Thanks, that's very interesting! There are indeed some things left we'd need to take care of. Not very nice without having a corresponding car stereo to test and benefit.

You can install android auto on your phone and test it (which I did) so it's not really a big issue.

> I wouldn't be too pessimistic here. What we offer with this PR is already on par with what YouTube does with voice search results: extremely recall oriented opportunistic search, with a high chance of false positives. So we can't learn from YouTube, but maybe from TuneIn.

Yes, I agree, I'm just trying to aim to the best end result =)

**Comment by morckx** [comment]
> You can install android auto on your phone and test it (which I did) so it's not really a big issue.

Great, I didn't now!

> Yes, I agree, I'm just trying to aim to the best end result =)

With one little addition voice search works perfectly (for my favourites only application).
Fallback to server search should be configurable, because precision definitely suffers if you're only interested in favorites.

**Pull request merged**